core/src/com/unciv/logic/map/tile/TileStatFunctions.kt

```diff
@@ -150,51 +173,82 @@ class TileStatFunctions(val tile: Tile) {
 
     // Only gets the tile percentage bonus, not the improvement percentage bonus
     @Suppress("MemberVisibilityCanBePrivate")
-    fun getTilePercentageStats(observingCiv: Civilization?, city: City?, uniqueCache: LocalUniqueCache): Stats {
-        val stats = Stats()
+    fun getTilePercentageStats(observingCiv: Civilization?, city: City?, uniqueCache: LocalUniqueCache): HashMap<String, Stats> {
+        val terrainStats = Stats()
         val stateForConditionals = StateForConditionals(civInfo = observingCiv, city = city, tile = tile)
 
+        val improvement = tile.getUnpillagedTileImprovement()
+        val improvementStats = Stats()
+
+        val road = if (!tile.roadIsPillaged && tile.roadStatus != RoadStatus.None)
+            tile.roadStatus.improvement(tile.ruleset)
+        else null
+        val roadStats = Stats()
+
         if (city != null) {
             val cachedStatPercentFromObjectCityUniques = uniqueCache.forCityGetMatchingUniques(
                 city, UniqueType.StatPercentFromObject, stateForConditionals)
 
             for (unique in cachedStatPercentFromObjectCityUniques) {
                 val tileFilter = unique.params[2]
-                if (tile.matchesTerrainFilter(tileFilter, observingCiv))
-                    stats[Stat.valueOf(unique.params[1])] += unique.params[0].toFloat()
+                if (tile.matchesFilter(tileFilter, observingCiv, true))
+                    terrainStats[Stat.valueOf(unique.params[1])] += unique.params[0].toFloat()
+                if (improvement != null && improvement.matchesFilter(tileFilter))
+                    improvementStats[Stat.valueOf(unique.params[1])] += unique.params[0].toFloat()
+                if (road != null && road.matchesFilter(tileFilter))
+                    roadStats[Stat.valueOf(unique.params[1])] += unique.params[0].toFloat()
```

Comment:
Let's use names. 
val amount = unique.params[0].toFloat()
val stat = Stat.valueOf(unique.params[1])

And then these if/elses